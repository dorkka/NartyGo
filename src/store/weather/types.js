export const SET_WEATHER = 'weather/SET_WEATHER';
export const IS_LOADING = 'weather/IS_LOADING';
export const SET_ERROR = 'weather/SET_ERROR';
