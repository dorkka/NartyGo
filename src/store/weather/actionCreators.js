import * as types from './types';
import resourceFetcher from '../../services/resourceFetcher';

export const setWeather = (data) => {
  const weatherTime = new Date().getTime;
  return {
    type: types.SET_WEATHER,
    payload: {
      data,
      weatherTime,
    },
  };
};
export const setError = (error) => ({
  type: types.SET_ERROR,
  error,
});

export const setIsLoading = () => ({
  type: types.IS_LOADING,
});

export const getWeather = (cityId) => (dispatch, getState) => {
  if ((getState().weather.weatherTime - (new Date().getTime)) < 7200000) {
    return;
  }
  dispatch(setIsLoading());
  resourceFetcher(`http://api.openweathermap.org/data/2.5/group?id=${cityId}&lang=pl&units=metric&APPID=07751527bfe01ed870793d6731eefdf1`)()
    .then(({ data }) => {
      dispatch(setWeather(data));
    })
    .catch(error => dispatch(setError(error)));
};
