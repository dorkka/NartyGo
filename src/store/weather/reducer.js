import * as types from './types';

const initialState = {
  byId: {},
  weatherTime: 0,
  isLoading: false,
  error: null,
};

function weatherReducer(state = initialState, action) {
  switch (action.type) {
    case types.SET_WEATHER: {
      return {
        ...state,
        byId: {
          ...state.byId,
          ...action.payload.byId,
        },
        weatherTime: action.meta.weatherTime,
        isLoading: false,
      };
    }
    case types.SET_ERROR: {
      return { ...state, error: action.error };
    }
    case types.IS_LOADING: {
      return { ...state, isLoading: true };
    }
    default: {
      return state;
    }
  }
}
export default weatherReducer;
