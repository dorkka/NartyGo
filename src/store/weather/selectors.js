import moment from 'moment';

export const getCurrentWeather = (state, cityId) => {
  const weatherFromStore = state.weather.byId[cityId] || {};
  const weatherWithDefaults = {
    main: {}, weather: [{}], wind: {}, sys: {}, ...weatherFromStore,
  };
  return ({
    temperature: weatherWithDefaults.main.temp,
    temperatureMin: weatherWithDefaults.main.temp_min,
    temperatureMax: weatherWithDefaults.main.temp_max,
    pressure: weatherWithDefaults.main.pressure,
    wind: weatherWithDefaults.wind.speed,
    clouds: weatherWithDefaults.weather[0].description,
    sunrise: moment(weatherWithDefaults.sys.sunrise).format('h:mm:ss'),
    sunset: moment(weatherWithDefaults.sys.sunset).format('h:mm:ss'),
  });
};

