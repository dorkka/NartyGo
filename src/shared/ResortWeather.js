import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ResortWeatherInfo from '../components/resortPage/ResortWeatherInfo';
import { getCurrentWeather } from '../store/weather/selectors';
import * as actions from '../store/weather/actionCreators';

export class ResortWeather extends Component {
  componentDidMount() {
    const { cityId } = this.props;
    this.props.getWeather(cityId);
  }

  render() {
    const { isLoading, error, currentWeather } = this.props;
    if (error) { return (error.message); }
    if (isLoading) { return <div>Weather Loading in progress</div>; }
    return (
      <div className="col-md-4">
        <ResortWeatherInfo {...currentWeather} />
      </div>
    );
  }
}

ResortWeather.propTypes = {
  getWeather: PropTypes.func.isRequired,
  currentWeather: PropTypes.object.isRequired,
  cityId: PropTypes.number.isRequired,
  isLoading: PropTypes.bool.isRequired,
  error: PropTypes.object.isRequired,
};

const mapStateToProps = (state, ownProps) => ({
  currentWeather: getCurrentWeather(state, ownProps.cityId),
  isLoading: state.weather.isLoading,
  error: state.weather.error,
});

const mapDispatchToProps = {
  getWeather: actions.getWeather,
};

export default connect(mapStateToProps, mapDispatchToProps)(ResortWeather);
