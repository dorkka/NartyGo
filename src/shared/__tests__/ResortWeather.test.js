import { ResortWeather } from '../ResortWeather';
import { currentWeather } from '../../specs/fixtures/weather';

describe('ResortWeather', () => {
  const defaultProps = {
    getWeather: () => { },
    currentWeather,
    cityId: 7531846,
    isLoading: false,
    error: null,
  };
  const setup = buildSetup(ResortWeather, defaultProps);

  describe('component', () => {
    test('should render component', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });
    test('should render error message', () => {
      const { wrapper } = setup({ error: { message: 'test error' } });
      expect(wrapper).toMatchSnapshot();
    });
    test('should render loading in progres when isLoading is truthy', () => {
      const { wrapper } = setup({ isLoading: true });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('componentDidMount', () => {
    test('should call getWeather with appropriate argument', () => {
      const { props } = setup({ getWeather: jest.fn() });
      expect(props.getWeather).toHaveBeenCalledWith(7531846);
    });
  });
});
