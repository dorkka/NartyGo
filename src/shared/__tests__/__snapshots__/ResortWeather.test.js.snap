// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResortWeather component should render component 1`] = `
<div
  className="col-md-4"
>
  <ResortWeatherInfo
    clouds="bezchmurnie"
    pressure={1020}
    sunrise="5:10:43"
    sunset="21:14:35"
    temperature={22}
    temperatureMax={30}
    temperatureMin={15}
    wind={1.5}
  />
</div>
`;

exports[`ResortWeather component should render error message 1`] = `"test error"`;

exports[`ResortWeather component should render loading in progres when isLoading is truthy 1`] = `
<div>
  Weather Loading in progress
</div>
`;
